daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -1732,10 +1736,74 @@ func (devices *DeviceSet) AddDevice(hash, baseHash string) error {
 		return fmt.Errorf("device %s already exists. Deleted=%v", hash, info.Deleted)
 	}
 
-	if err := devices.createRegisterSnapDevice(hash, baseInfo); err != nil {
+	if err := devices.createRegisterSnapDevice(hash, baseInfo, size); err != nil {
 		return err
 	}
 
+	// Grow the container rootfs.
+	if size > 0 {
+		if err := devices.growFS(hash); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (devices *DeviceSet) growFS(hash string) error {
+
+	info, err := devices.lookupDevice(hash)
+	if info == nil {
+		return fmt.Errorf("Failed to lookup device %s:%v", hash, err)
+	}
+
+	info.lock.Lock()
+	defer info.lock.Unlock()
+
+	if err := devices.activateDeviceIfNeeded(info, false); err != nil {
+		return fmt.Errorf("Error activating devmapper device for '%s': %s", hash, err)
+	}
+
+	defer devices.deactivateDevice(info)
+
+	fstype, err := ProbeFsType(info.DevName())
```

Comment:
this is now stored in `devices.BaseDeviceFilesystem` (need latest docker/docker master code)